### PR TITLE
feat(messaging): support loc keys on notification payloads

### LINKED ISSUE #3509: 🔥 RemoteMessage does not expose localization keys from native message
## Issue

When a Firebase message is sent with localized strings (body_loc_key/body_loc_args and title_loc_key/title_loc_args on Android, loc-key/loc-args and title-loc-key and title-loc-args on iOS), there is no way to get at these values from TypeScript.

This is a blocker for displaying local notifications when the app is foregrounded - when in the background the OS does the right thing, but I need these values exposed to the script layer in order to properly localize the manually constructed notification.

This should be fixable in `ReactNativeFirebaseMessagingSerializer` and `RNFBMessagingSerializer` - the native APIs should provide everything needed to populate these strings/arrays when serializing.

On Android at least I have verified this can be addressed in the serializer by calling `notification.getTitleLocalizationKey`/`getTitleLocalizationArgs`/`getBodyLocalizationKey`/`getBodyLocalizationArgs` and serializing as appropriate. I would be happy to contribute a PR (at least for Android, not set up to test this scenario on iOS at the moment).

Sorry - I was not sure this would count as a feature request vs a bug since it involves additions to a type/interface, but it struck me as a gap in the serializer that caused fundamental difficulty when using the library.

---

### Javascript

<details><summary>Click To Expand</summary>
<p>

#### `package.json`:

```json
    "@react-native-firebase/app": "~6.4.0",
    "@react-native-firebase/messaging": "~6.4.0",
```

#### `firebase.json` for react-native-firebase v6:

```json
# N/A
```

</details>

### iOS

<details><summary>Click To Expand</summary>
<p>

#### `ios/Podfile`:

- [ ] I'm not using Pods
- [x] I'm using Pods and my Podfile looks like:

```ruby
# N/A
```

#### `AppDelegate.m`:

```objc
// N/A
```

</p>
</details>

---

### Android

<details><summary>Click To Expand</summary>
<p>

#### Have you converted to AndroidX?

- [ ] my application is an AndroidX application?
- [ ] I am using `android/gradle.settings` `jetifier=true` for Android compatibility?
- [x] I am using the NPM package `jetifier` for react-native compatibility?

#### `android/build.gradle`:

```groovy
// N/A
```

#### `android/app/build.gradle`:

```groovy
// N/A
```

#### `android/settings.gradle`:

```groovy
// N/A
```

#### `MainApplication.java`:

```java
// N/A
```

#### `AndroidManifest.xml`:

```xml
<!-- N/A -->
```

</p>
</details>

---

## Environment

<details><summary>Click To Expand</summary>
<p>

**`react-native info` output:**

```
System:
    OS: Windows 10 10.0.18363
    CPU: (8) x64 Intel(R) Core(TM) i7-7700K CPU @ 4.20GHz
    Memory: 1.27 GB / 15.95 GB
  Binaries:
    Node: 10.13.0 - C:\Program Files\nodejs\node.EXE
    Yarn: 1.22.4 - C:\Users\xxx\AppData\Roaming\npm\yarn.CMD
    npm: 6.4.1 - C:\Program Files\nodejs\npm.CMD
  IDEs:
    Android Studio: Version  3.1.0.0 AI-173.4907809
  npmPackages:
    react: 16.9.0 => 16.9.0
    react-native: 0.61.5 => 0.61.5
```

- **Platform that you're experiencing the issue on**:
  - [ ] iOS
  - [ ] Android
  - [ ] **iOS** but have not tested behavior on Android
  - [x] **Android** but have not tested behavior on iOS
  - [ ] Both
- **`react-native-firebase` version you're using that has this issue:**
  - `6.4.0`
- **`Firebase` module(s) you're using that has the issue:**
  - `messaging`
- **Are you using `TypeScript`?**
  - `Y: 3.7.2`

</p>
</details>

---

- 👉 Check out [`React Native Firebase`](https://twitter.com/rnfirebase) and [`Invertase`](https://twitter.com/invertaseio) on Twitter for updates on the library.

### PR DESCRIPTION
**Pull request by sfuqua**
### Description

`RemoteMessage.notification` does not currently expose localization keys if the service expects clients to localize messages.

Exposing these fields to Typescript if they exist will be beneficial, because it allows the TS part of the client to localize a local notification based on the same loc keys and formatting that they already have for the native notifications in the app.

This change adds support for the following fields on both platforms:
* titleLocKey: string | undefined
* titleLocArgs: string[] | undefined
* bodyLocKey: string | undefined
* bodyLocArgs: string[] | undefined

I also added support for subtitleLocKey and subtitleLocArgs on iOS.

* [Firebase docs that include these fields with the same names](https://firebase.google.com/docs/reference/admin/node/admin.messaging.AndroidNotification)
* [Firebase/Google docs that include the relevant Java methods for RemoteMessage.Notification](https://firebase.google.com/docs/reference/android/com/google/firebase/messaging/RemoteMessage.Notification)
* [iOS documentation that includes relevant fields](https://developer.apple.com/documentation/usernotifications/setting_up_a_remote_notification_server/generating_a_remote_notification#2943365)

### Related issues

Fixes #3509 

### Release Summary

Exposes title/body localization keys and arguments in RemoteMessage.notification.

### Checklist

- I read the [Contributor Guide](/CONTRIBUTING.md) and followed the process outlined there for submitting PRs.
  - [x] Yes
- My change supports the following platforms;
  - [x] `Android`
  - [x] `iOS`
- My change includes tests;
  - [ ] `e2e` tests added or updated in `packages/\*\*/e2e`
  - [ ] `jest` tests added or updated in `packages/\*\*/__tests__`
- [x] I have updated TypeScript types that are affected by my change.
- This is a breaking change;
  - [ ] Yes
  - [x] No

### Test Plan

I have verified that the code works as expected on Android using a private package in my own application, with Firebase messages that include these fields.

Will need help testing iOS - I verified the code builds but have no way of validating FCM with an iOS client right now.

---

Think `react-native-firebase` is great? Please consider supporting the project with any of the below:

- 👉 Star this repo on GitHub ⭐️
- 👉 Follow [`React Native Firebase`](https://twitter.com/rnfirebase) and [`Invertase`](https://twitter.com/invertaseio) on Twitter

:fire:

### CODE REVIEW

**Comment by Salakar** [comment]
@sfuqua thanks for this well put together PR :star: would you mind signing the CLA please?

We'd be happy to accept this as a feature as I think this should be supported. cc @russellwheatley @dackers86 who should be able to help verify this on iOS.

**Comment by sfuqua** [comment]
> @sfuqua thanks for this well put together PR ⭐ would you mind signing the CLA please?
> 
> We'd be happy to accept this as a feature as I think this should be supported. cc @russellwheatley @dackers86 who should be able to help verify this on iOS.

Cheers @Salakar! I am waiting on an internal review before I'm able to sign the CLA, as this contribution was done as part of my work with Microsoft and I need a green light before proceeding. I will provide an update as soon as I can - it looks like there was movement earlier today.

**Comment by Salakar** [comment]
> > @sfuqua thanks for this well put together PR star would you mind signing the CLA please?
> > We'd be happy to accept this as a feature as I think this should be supported. cc @russellwheatley @dackers86 who should be able to help verify this on iOS.
> 
> Cheers @Salakar! I am waiting on an internal review before I'm able to sign the CLA, as this contribution was done as part of my work with Microsoft and I need a green light before proceeding. I will provide an update as soon as I can - it looks like there was movement earlier today.

Ok cool, let me know if any issues, thanks

**Comment by sfuqua** [comment]
I just signed the CLA.

**Comment by sfuqua** [comment]
@Salakar fyi in case you missed the CLA update last week (apologies otherwise) - I am no longer blocked on this one. Let me know if I can do anything else to help on my end! :)

File: packages/messaging/ios/RNFBMessaging/RNFBMessagingSerializer.m [hunk 1]
```diff
@@ -130,18 +130,50 @@ + (NSDictionary *)remoteMessageUserInfoToDict:(NSDictionary *)userInfo {
...
         // message.notification.ios.subtitle
         if (apsAlertDict[@"subtitle"] != nil) {
           notificationIOS[@"subtitle"] = apsAlertDict[@"subtitle"];
         }
-
+          
+        // iOS only
+        // message.notification.ios.subtitleLocKey
+        if (apsAlertDict[@"subtitle-loc-key"] != nil) {
+          notification[@"subtitleLocKey"] = apsAlertDict[@"subtitle-loc-key"];
```
Comment (Salakar) [inline packages/messaging/ios/RNFBMessaging/RNFBMessagingSerializer.m:168]:
```suggestion
          notificationIOS[@"subtitleLocKey"] = apsAlertDict[@"subtitle-loc-key"];
```

File: packages/messaging/ios/RNFBMessaging/RNFBMessagingSerializer.m [hunk 2]
```diff
@@ -130,18 +130,50 @@ + (NSDictionary *)remoteMessageUserInfoToDict:(NSDictionary *)userInfo {
...
+        }
+          
+        // iOS only
+        // message.notification.ios.subtitleLocArgs
+        if (apsAlertDict[@"subtitle-loc-args"] != nil) {
+          notification[@"subtitleLocArgs"] = apsAlertDict[@"subtitle-loc-args"];
```
Comment (Salakar) [inline packages/messaging/ios/RNFBMessaging/RNFBMessagingSerializer.m:174]:
```suggestion
          notificationIOS[@"subtitleLocArgs"] = apsAlertDict[@"subtitle-loc-args"];
```

**Comment by Salakar** [comment]
Heya, v7.0.0 release is up with this change: https://github.com/invertase/react-native-firebase/blob/master/packages/messaging/CHANGELOG.md

The major version bump is just to signify some internal versioning changes we've done - there's no breaking changes - can just update directly to v7 for this 👍

**Pull request merged**